refactor(web-components): refactor menu to popover

davatron5000 opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [x] Code is up-to-date with the `master` branch
* [x] Your changes are covered by tests (if possible)
* [ ] You've run `yarn change` locally


PR flow tips:
* [x] Try to start with a Draft PR
* [ ] Once you're ready (ideally the pipeline is passing) promote your PR to Ready for Review. This step will auto-assign reviewers for your PR.
-->

## Previous Behavior

<!-- This is the behavior we have today -->

The current Menu component has a hard dependency on `@floating-ui/dom`, this is an attempt to slim that down and leverage HTML `popover` and CSS `anchor` positioning. 

Also discovered: 

- Drop-shadow bug on nested menu items
- Nested menu items not working in Safari

## New Behavior

<!-- This is the behavior we should expect with the changes in this PR -->

This component has been refactored to use HTML `popover` and CSS `anchor` positioning. HTML `popover`is polyfillable and we're able to gracefully degrade CSS `anchor`, so that's promising. 

- [x] Refactor `Menu` to use the HTML `[popover]` and CSS `anchor`
- [x] Allow for setting `max-height` of dropdown (per request from Commerce UI Kit) 
- [x] Refactor `MenuList` and `MenuItem` to use `[popover]` as well.
- [x] Add default fallback position for browsers that don't support CSS anchor
- [x] Remove `@floating-ui/dom` dependency from `Menu` and `MenuItem`

Due to platform limitations with popover, we're unable to apply `popovertarget` to `fluent-button` (because it's not a _real_ `<button>`) which would automate some of the show/hiding focus management and ARIA-expanded work, but have included workarounds.

## Known issues and TODOs

- [x] `:focus-visible` not visible on first and last items in Safari
- [x] `contextmenu` is a little futzy
- [ ] Document HTML popover polyfill
- [x] Run `yarn change`
- [x] Update API Extractor
- [x] Update `package.json`

## Related Issue(s)

<!-- Please link the issue being fixed so it gets closed when this is merged. -->

- Fixes #

fabricteam commented:
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

fabricteam commented on packages/web-components/src/menu/menu.styles.ts:
```diff
@@ -1,18 +1,36 @@
 import { css } from '@microsoft/fast-element';
```
## [🕵 fluentui-web-components-v3 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31674&commitId=1194671ea15ff379ba4a06608ecb98eb4a579113&env=prod&clientType=FLUENTUI)


  <details><summary><b>Accordion</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Accordion.Size - Dark Mode.normal.chromium.png|2| Changed|
|Accordion.Size.normal.chromium.png|2| Changed|
|Accordion.Size - RTL.normal.chromium.png|2| Changed|

</details>


  <details><summary><b>Avatar</b> 5 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar.Active Appearance - Dark Mode.normal.chromium.png|2| Changed|
|Avatar.Active.normal.chromium.png|5| Changed|
|Avatar.Active Appearance.normal.chromium.png|10| Changed|
|Avatar.Color.normal.chromium.png|1| Changed|
|Avatar.Color - RTL.normal.chromium.png|1| Changed|

</details>


  <details><summary><b>Badge</b> 6 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Badge.Appearance - Dark Mode.normal.chromium.png|15| Changed|
|Badge.Appearance.normal.chromium.png|1| Changed|
|Badge.Default.normal.chromium.png|1| Changed|
|Badge.Shape.normal.chromium.png|1| Changed|
|Badge.Size.normal.chromium.png|6| Changed|
|Badge.Color.normal.chromium.png|1| Changed|

</details>


  <details><summary><b>Button</b> 57 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Button.Large Icon Only - Dark Mode.default.chromium.png|4| Changed|
|Button.Large Icon Only.pressed.chromium.png|3| Changed|
|Button.Large Icon Only - Dark Mode.pressed.chromium.png|5| Changed|
|Button.Large With Icon - Dark Mode.hover.chromium.png|3| Changed|
|Button.Large With Icon - RTL.hover.chromium.png|4| Changed|
|Button.Large With Icon - Dark Mode.pressed.chromium.png|5| Changed|
|Button.Large Icon Only.default.chromium.png|5| Changed|
|Button.Large With Icon - RTL.pressed.chromium.png|3| Changed|
|Button.Large Icon Only - Dark Mode.hover.chromium.png|3| Changed|
|Button.Large With Icon.pressed.chromium.png|3| Changed|
|Button.Large With Icon.hover.chromium.png|4| Changed|
|Button.Large Icon Only.hover.chromium.png|4| Changed|
|Button.Large With Icon.default.chromium.png|5| Changed|
|Button.Large With Icon - Dark Mode.default.chromium.png|4| Changed|
|Button.Large With Icon - RTL.default.chromium.png|5| Changed|
|Button.Outline Disabled - Dark Mode.pressed.chromium.png|159| Changed|
|Button.Outline Disabled - Dark Mode.default.chromium.png|159| Changed|
|Button.Primary Disabled - Dark Mode.default.chromium.png|44| Changed|
|Button.Outline Disabled Focusable - Dark Mode.hover.chromium.png|159| Changed|
|Button.Primary Disabled Focusable.hover.chromium.png|15| Changed|
|Button.Primary Disabled Focusable - Dark Mode.pressed.chromium.png|44| Changed|
|Button.Primary Disabled Focusable - Dark Mode.default.chromium.png|44| Changed|
|Button.Outline Disabled Focusable - Dark Mode.pressed.chromium.png|159| Changed|
|Button.Primary Disabled Focusable.default.chromium.png|15| Changed|
|Button.Outline Disabled Focusable - Dark Mode.default.chromium.png|159| Changed|
|Button.Primary Disabled Focusable.pressed.chromium.png|15| Changed|
|Button.Primary Disabled Focusable - Dark Mode.hover.chromium.png|44| Changed|
|Button.Primary Disabled - Dark Mode.pressed.chromium.png|44| Changed|
|Button.Primary Disabled.hover.chromium.png|15| Changed|
|Button.Outline Disabled - Dark Mode.hover.chromium.png|159| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.default.chromium.png|187| Changed|
|Button.Subtle Disabled - Dark Mode.pressed.chromium.png|187| Changed|
|Button.Subtle Disabled - Dark Mode.default.chromium.png|187| Changed|
|Button.Primary Disabled.pressed.chromium.png|15| Changed|
|Button.Subtle Disabled.pressed.chromium.png|25| Changed|
|Button.Primary Disabled.default.chromium.png|15| Changed|
|Button.Subtle Disabled.hover.chromium.png|25| Changed|
|Button.Transparent Disabled - Dark Mode.default.chromium.png|199| Changed|
|Button.Subtle Disabled.default.chromium.png|25| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.hover.chromium.png|199| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.default.chromium.png|199| Changed|
|Button.Transparent Disabled.hover.chromium.png|27| Changed|
|Button.Transparent Disabled.pressed.chromium.png|27| Changed|
|Button.Subtle Disabled Focusable.default.chromium.png|25| Changed|
|Button.Transparent Disabled Focusable.hover.chromium.png|27| Changed|
|Button.Subtle Disabled - Dark Mode.hover.chromium.png|187| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.hover.chromium.png|187| Changed|
|Button.Transparent Disabled Focusable.pressed.chromium.png|27| Changed|
|Button.Transparent Disabled - Dark Mode.pressed.chromium.png|199| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.pressed.chromium.png|199| Changed|
|Button.Subtle Disabled Focusable.pressed.chromium.png|25| Changed|
|Button.Primary Disabled - Dark Mode.hover.chromium.png|44| Changed|
|Button.Transparent Disabled.default.chromium.png|27| Changed|
|Button.Transparent Disabled - Dark Mode.hover.chromium.png|199| Changed|
|Button.Transparent Disabled Focusable.default.chromium.png|27| Changed|
|Button.Subtle Disabled Focusable.hover.chromium.png|25| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.pressed.chromium.png|187| Changed|

</details>


  <details><summary><b>Checkbox</b> 26 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Checkbox.Checked - Dark Mode.normal.chromium.png|982| Changed|
|Checkbox.Circular Checked.normal.chromium.png|4883| Changed|
|Checkbox.Circular Indeterminate.indeterminate.chromium.png|4869| Changed|
|Checkbox.Checked.normal.chromium.png|4855| Changed|
|Checkbox.Circular Indeterminate - Dark Mode.indeterminate.chromium.png|1102| Changed|
|Checkbox.Default.normal.chromium.png|4751| Changed|
|Checkbox.Default - Dark Mode.normal.chromium.png|814| Changed|
|Checkbox.Circular Indeterminate - Dark Mode.normal.chromium.png|936| Changed|
|Checkbox.Default - RTL.normal.chromium.png|4751| Changed|
|Checkbox.Disabled Checked.normal.chromium.png|4721| Changed|
|Checkbox.Indeterminate - Dark Mode.normal.chromium.png|877| Changed|
|Checkbox.Indeterminate - Dark Mode.indeterminate.chromium.png|993| Changed|
|Checkbox.Disabled.normal.chromium.png|4705| Changed|
|Checkbox.Circular Checked - Dark Mode.normal.chromium.png|1080| Changed|
|Checkbox.Circular Disabled.normal.chromium.png|4711| Changed|
|Checkbox.Circular Disabled - Dark Mode.normal.chromium.png|760| Changed|
|Checkbox.Indeterminate.indeterminate.chromium.png|4847| Changed|
|Checkbox.Circular - Dark Mode.normal.chromium.png|792| Changed|
|Checkbox.Circular Disabled Checked.normal.chromium.png|4728| Changed|
|Checkbox.Disabled Checked - Dark Mode.normal.chromium.png|777| Changed|
|Checkbox.Indeterminate.normal.chromium.png|4776| Changed|
|Checkbox.Circular Disabled Checked - Dark Mode.normal.chromium.png|799| Changed|
|Checkbox.Circular.normal.chromium.png|4734| Changed|
|Checkbox.Disabled - Dark Mode.normal.chromium.png|739| Changed|
|Checkbox.Circular - RTL.normal.chromium.png|4733| Changed|
|Checkbox.Circular Indeterminate.normal.chromium.png|4786| Changed|

</details>


  <details><summary><b>Divider</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Divider.With Svg.normal.chromium.png|3| Changed|
|Divider.With Svg - RTL.normal.chromium.png|3| Changed|
|Divider.Vertical With Svg.normal.chromium.png|4| Changed|

</details>


  <details><summary><b>Label</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Label.Size Small.normal.chromium.png|661| Changed|
|Label.Required - RTL.normal.chromium.png|11| Changed|

</details>


  <details><summary><b>MenuList</b> 16 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|MenuList.Checkbox With Icons.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox With Icons.2nd selected.chromium.png|5| Changed|
|MenuList.Checkbox With Icons - RTL.2nd selected.chromium.png|5| Changed|
|MenuList.Checkbox With Icons - RTL.normal.chromium.png|4| Changed|
|MenuList.Checkbox With Icons - RTL.1st selected.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox.2nd selected.chromium.png|1| Changed|
|MenuList.Radio With Icons.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox - RTL.2nd selected.chromium.png|1| Changed|
|MenuList.Radio With Icons.normal.chromium.png|4| Changed|
|MenuList.Checkbox With Icons.normal.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.2nd selected.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.normal.chromium.png|4| Changed|
|MenuList.With Icons.hover menuitem.chromium.png|3| Changed|
|MenuList.With Icons.normal.chromium.png|4| Changed|
|MenuList.Radio With Icons.2nd selected.chromium.png|4| Changed|

</details>


  <details><summary><b>Slider</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Slider.Size Small.normal.chromium.png|1| Changed|
|Slider.Size Small.rightArrow.chromium.png|1| Changed|

</details>


  <details><summary><b>Switch</b> 34 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Switch.Checked - Dark Mode.normal.chromium.png|1130| Changed|
|Switch.Default - Dark Mode.hover.chromium.png|1071| Changed|
|Switch.Disabled - Dark Mode.normal.chromium.png|909| Changed|
|Switch.Checked.normal.chromium.png|5132| Changed|
|Switch.Disabled Checked.hover.chromium.png|5047| Changed|
|Switch.Disabled.hover.chromium.png|4914| Changed|
|Switch.Disabled - RTL.normal.chromium.png|4914| Changed|
|Switch.Checked - Dark Mode.hover.chromium.png|1169| Changed|
|Switch.Disabled.normal.chromium.png|4914| Changed|
|Switch.Checked - RTL.normal.chromium.png|5132| Changed|
|Switch.Disabled - Dark Mode.hover.chromium.png|909| Changed|
|Switch.Checked - RTL.hover.chromium.png|5150| Changed|
|Switch.Default - Dark Mode.normal.chromium.png|1019| Changed|
|Switch.Checked.hover.chromium.png|5150| Changed|
|Switch.Default - RTL.hover.chromium.png|5044| Changed|
|Switch.Disabled - RTL.hover.chromium.png|4914| Changed|
|Switch.Default.normal.chromium.png|5009| Changed|
|Switch.Default - RTL.normal.chromium.png|5009| Changed|
|Switch.Default.hover.chromium.png|5044| Changed|
|Switch.Disabled Checked - Dark Mode.normal.chromium.png|1074| Changed|
|Switch.Disabled Checked.normal.chromium.png|5047| Changed|
|Switch.Disabled Checked - Dark Mode.hover.chromium.png|1074| Changed|
|Switch.Label Above - RTL.hover.chromium.png|0| Removed|
|Switch.Label Above - RTL.normal.chromium.png|0| Removed|
|Switch.Label Above.hover.chromium.png|0| Removed|
|Switch.Label Above.normal.chromium.png|0| Removed|
|Switch.Label After - RTL.hover.chromium.png|0| Removed|
|Switch.Label After - RTL.normal.chromium.png|0| Removed|
|Switch.Label After.hover.chromium.png|0| Removed|
|Switch.Label After.normal.chromium.png|0| Removed|
|Switch.Label Before - RTL.hover.chromium.png|0| Removed|
|Switch.Label Before - RTL.normal.chromium.png|0| Removed|
|Switch.Label Before.hover.chromium.png|0| Removed|
|Switch.Label Before.normal.chromium.png|0| Removed|

</details>


  <details><summary><b>Text</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Text.Strikethrough.normal.chromium.png|236| Changed|
|Text.Block.normal.chromium.png|995| Changed|

</details>


  <details><summary><b>TextInput</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|TextInput.Size Large.normal.chromium.png|3| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_web-components"/>

davatron5000 replied:
I confirmed that all the MenuList related screenshots were noise or should be accepted. Unclear on how to do partial approvals from the VR Tool UI. So going to resolve this conversation and we can fix VR tests in a future PR.

marchbox commented on packages/web-components/src/menu-item/menu-item.ts:
```diff
@@ -183,23 +130,20 @@ export class MenuItem extends FASTElement {
-        this.expanded && this.submenu ? this.submenu.focus() : this.expandAndFocus();
+        if (this.disabled) return false;
+        this.submenu?.togglePopover(true);
+        this.submenu?.focus();
```
To save one duplicated `return false`, maybe:

```js
if (!this.disabled) {
  this.submenu?.togglePopover(true);
  this.submenu?.focus();
}
return false;
```

marchbox commented on packages/web-components/src/menu-item/menu-item.ts:
```diff
@@ -183,23 +130,20 @@ export class MenuItem extends FASTElement {
+        try {
+          this.parentElement?.togglePopover(false);
+        } catch (e) {
+          // Catch DOMException if parentElement is not a menu
```
Maybe check if `parentElement` is a menu instead of using try...catch? I think it'd make the code more readable and explicit.

chrisdholt commented on packages/web-components/src/menu-item/menu-item.template.ts:
```diff
@@ -14,69 +14,34 @@ const chevronRight16Filled = html.partial(
-          )}
+      @toggle="${(x, c) => x.toggleHandler(c.event as ToggleEvent)}"
+    >
+      <div part="indicator" class="indicator">
+        <slot name="indicator"> ${staticallyCompose(options.indicator)} </slot>
       </div>
```
Question/thought - Can we make this and `submenu-glyph` just a slot and then add the corresponding classes to the default SVG's?

chrisdholt replied:
Essentially all we'd have here is `<slot name="indicator"> ${staticallyCompose(options.indicator)} </slot>`

davatron5000 replied:
When I tried this, it all fell apart because the grid really depended on the structure. BUT. I'll try again.

chrisdholt replied:
I pulled this down and tried it - but I *might* have missed a configuration perhaps that we don't have in the storybook. :)

davatron5000 replied:
@chrisdholt I think I got it working with flatter DOM, but just so I'm clear, this is what you were thinking? 

```js
const Checkmark16Filled = html.partial(
  `<svg class="indicator" fill="currentColor" ...`,
);
const chevronRight16Filled = html.partial(
  `<svg class="submenu-glyph" fill="currentColor" ...`,
);
```

Then would consumers would then need that `.indicator` or `.submenu-glyph` class on their slotted SVGs? 
Or should I also do `::slotted([slot="indicator"]), ::slotted([slot="submenu-glyph]])`  for styling and positioning slotted content?

chrisdholt replied:
I think the selectors can stay as they are - the classes in the stylesheet would map to the default slotted content whereas the `::slotted()` selectors would be for anything coming in the light DOM.

davatron5000 replied:
Okay! I think I got it all flattened. LMK what you think @chrisdholt

chrisdholt replied:
Excellent!

chrisdholt commented on packages/web-components/src/menu-item/menu-item.styles.ts:
```diff
@@ -41,46 +42,49 @@ export const styles = css`
-  :host([checked]) .radio {
+  ::slotted([slot='end']) {
     display: inline-flex;
     justify-content: center;
     align-items: center;
```
I think we can drop these if we flatten the DOM for the indicator/submenu-glyph.

chrisdholt commented on packages/web-components/src/menu-item/menu-item.styles.ts:
```diff
@@ -91,13 +95,12 @@ export const styles = css`
 
-  :host(:hover) .input-container,
-  :host(:hover) .expand-collapse-glyph-container,
+  :host(:hover) .indicator,
+  :host(:hover) .submenu-glyph,
   :host(:hover) .content {
```
Can we leverage `:is()` here?

chrisdholt commented on packages/web-components/src/menu-item/menu-item.styles.ts:
```diff
@@ -110,8 +113,8 @@ export const styles = css`
 
-  :host(:active) .input-container,
-  :host(:active) .expand-collapse-glyph-container,
+  :host(:active) .indicator,
+  :host(:active) .submenu-glyph,
   :host(:active) .content {
```
Can we leverage `:is()` here as well (and once more below)?

chrisdholt commented on packages/web-components/src/menu-item/menu-item.styles.ts:
```diff
@@ -41,46 +42,49 @@ export const styles = css`
+  .submenu-glyph,
   ::slotted([slot='start']),
   ::slotted([slot='end']) {
     height: 32px;
```
if we flatten the DOM I *think* this can go away?

chrisdholt reviewed (commented):
This looks great! One place where we'll want to remove the `@floating-ui/dom` dependency as well is the tensile config for the benchmarks. There's a reference in there we can likely remove.

mlijanto commented on packages/web-components/src/menu-item/menu-item.ts:
```diff
@@ -236,47 +180,37 @@ export class MenuItem extends FASTElement {
     }
 
-    this.expanded = false;
+    this.submenu?.togglePopover(false);
```
This is more of a design question. Do we want to add a delay before closing the submenu? The delay will allow the cursor to cross from the menu item to a long submenu if needed.

chrisdholt commented on packages/web-components/src/menu/menu.styles.ts:
```diff
@@ -1,18 +1,35 @@
   :host {
-    position: relative;
-    z-index: var(--z-index-menu, 1);
+    display: inline-block;
```
I think this one can be removed :)

mlijanto commented on packages/web-components/src/menu/menu.styles.ts:
```diff
@@ -1,18 +1,35 @@
+  ::slotted([popover]) {
+    inset-area: block-end span-inline-end;
+    margin: 0;
+    max-height: var(--menu-max-height, auto);
```
It would be great to document this somewhere. Maybe add a note to include this when we update the docs?

davatron5000 replied:
Let me make you a story for that. And I'll note the CSS property on the class.

davatron5000 replied:
Documented in `menu.ts` and Storybook and I found a bug and fixed that.

mlijanto replied:
Thanks! And thanks for adding this to menu!

mlijanto commented:
Looks like there's a couple of color bugs in menu-item when disabled. Here's what I see:

Disabled, rest - looks good!
<img width="171" alt="image" src="https://github.com/microsoft/fluentui/assets/601470/f77178ab-e70f-4fbc-a9c9-d51375f1427f">

Disabled, active - icon color should not change color?
<img width="167" alt="image" src="https://github.com/microsoft/fluentui/assets/601470/fea8f91e-7cf6-4d8f-9ee2-8b30ba4ed568">

Disabled, high contrast - no color difference, maybe needs a system color override
<img width="170" alt="image" src="https://github.com/microsoft/fluentui/assets/601470/b2460266-753f-4a8f-bfd9-53c471475450">

radium-v commented:
A small interaction nit with menu: When clicking a menu to open it, keyboard navigation doesn't work. You have to redirect focus away from the control then back into it to use the keyboard. This isn't an issue with menu-list.

radium-v commented on packages/web-components/src/menu-item/menu-item.ts:
```diff
@@ -125,48 +88,27 @@ export class MenuItem extends FASTElement {
     if (next.length) {
       this.submenu = next[0];
-      this.updateSubmenu();
+      this.submenu.setAttribute('popover', '');
```
nit: use `toggleAttribute` for boolean attributes:

```suggestion
      this.submenu.toggleAttribute('popover', true);
```

radium-v commented on packages/web-components/src/menu-item/menu-item.ts:
```diff
@@ -217,66 +156,43 @@ export class MenuItem extends FASTElement {
-    }
-
-    this.expanded = true;
+    if (this.disabled) return false;
```
```suggestion
    if (this.disabled) {
        return false;
    }
```

radium-v commented on packages/web-components/src/menu-item/menu-item.ts:
```diff
@@ -310,46 +227,30 @@ export class MenuItem extends FASTElement {
+        // If an open submenu is too wide for the viewport, move it above.
+        this.submenu.style.translate = '0 -100%';
+        return;
+      } else if (thisRect[inlineEnd] + thisSubmenuRect.width > window.innerWidth) {
```
The previous `if` returns, so this doesn't need an `else`

davatron5000 commented:
@mlijanto Thanks for finding that. 🙌 

- [x] Fixed the `[disabled]:active` issue. 
- [x] Added a `disabled` menuitem to the story as well.
- [x] Fixed `disabled` state in forced-colors mode.

@radium-v 

- [x] Rolled in your nits
- [x] Fixed the focus situation on `Menu` when you click and then want to use arrow keys.
